Vulkan AS rebuild-on-replay: Serialise and Replay

cmannett85-arm opened the pull request:
The `VkAccelerationStructureInfo` structure is serialised followed by the unified geometry readback buffer.  On replay the upload buffer is populated from the capture and then copied to GPU local mem, at which point the upload mem is freed as replay cannot modify the AS input data.

When the initial state Apply() is called the ASes are built from the input data one at a time so that a single scratch buffer can be built (enlarging when needed) and re-used.  Although this is slower it is necessary on Mali as it has poor space efficiency for the scratch buffer and so can easily OOM a device if all the ASes are built in a single command buffer submission.

Tested on:

- Android Mali G715 r49p1 with various Arm tech demos (some based on UE5) and Vulkan Samples ray_queries
- Ubuntu 24.04 NV 3090FE 535.183.06, same above but ported to x86_64

cmannett85-arm committed:
Vulkan AS rebuild-on-replay: Serialise and Replay

The VkAccelerationStructureInfo structure is serialised followed by the unified geometry readback buffer.

When the initial state Apply() is called the ASes are built from the input data one at a time so that a single scratch buffer can be built (enlarging when needed) and re-used.  Although this is slower it is necessary on Mali as it has poor space efficiency for the scratch buffer and so can easily OOM a device if all the ASes are built in a single command buffer submission.

baldurk commented on renderdoc/driver/vulkan/vk_manager.h:
```diff
@@ -113,7 +114,10 @@ struct VkInitialContents
 
-    // MemoryAllocation and serialised ASes are not free'd here
+    if(accelerationStructureInfo)
+      accelerationStructureInfo->Release();
```
This can be `SAFE_RELEASE` which does this for you.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -51,6 +51,77 @@ VkDeviceSize IndexTypeSize(VkIndexType type)
+{
+  const uint64_t aabbsSize = sizeof(VkDeviceSize);    // stride
+
+  // You can't just use sizeof(Triangles) due to padding
```
I think you could use `sizeof()` here to reduce the calculation for this? The size returned by `GetSize_InitialState` only has to be an upper bound so we can write the chunk header as we go without rewinding to fixup. As long as it's not a hugely pessimistic upper bound a little bit of padding per chunk is fine and will be dominated by the data itself.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -68,6 +139,103 @@ void VkAccelerationStructureInfo::Release()
+        VkDeviceOrHostAddressConstKHR tData;
+        tData.deviceAddress = g.tris.hasTransformData ? 0x1 : 0x0;
+
+        geoUnion.triangles = VkAccelerationStructureGeometryTrianglesDataKHR{
```
Please do not use this syntax. The typename can be omitted here if you're initialising POD with `{}` but either use a normal constructor or a POD brace list.

baldurk replied:
This line is still using the odd construction syntax.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -450,320 +618,261 @@ void VulkanAccelerationStructureManager::CopyAccelerationStructure(
-    RDCERR("Couldn't acquire command buffer");
-    return false;
-  }
+bool VulkanAccelerationStructureManager::Serialise(WriteSerialiser &ser, ResourceId id,
```
Please keep these serialise functions as templated over the serialiser type rather than explicitly specialising for read and write. I know in cases like this with initial contents there is not much commonality but every other serialise function is written like this, with conditional code for each mode, so I would not want to have an exception only for this case.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -450,320 +618,261 @@ void VulkanAccelerationStructureManager::CopyAccelerationStructure(
-      const VkDeviceSize size = AlignUp(initial->mem.size, nonCoherentAtomSize);
+      uploadMemory =
+          CreateReplayMemory(MemoryType::Upload, geomData.memSize,
+                             VK_BUFFER_USAGE_ACCELERATION_STRUCTURE_BUILD_INPUT_READ_ONLY_BIT_KHR);
```
The upload memory does not need this build input usage flag.

baldurk replied:
I see that you're now building the AS directly from this buffer so it does need the flag now, though I'm a little worried about the perf impact of doing a build directly from this memory rather than the intermediate GPU-local buffer you had before in the optimal memory type.

Was there a specific reason you removed that? It would still be destroyed at the same time as the upload buffer after the first build & cache but would avoid any problems where the build ends up running slow because it's reading from upload memory.

cmannett85-arm replied:
GPU memory is usually the limiting factor on mobile so I just went with one copy, if performance is the bigger concern on desktop then I'll switch to that and maybe keep this path in our fork.

baldurk replied:
We only need to hold onto the GPU-local buffer for a short period of time so yes please change this to take the performant path.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -836,6 +945,162 @@ VulkanAccelerationStructureManager::Allocation VulkanAccelerationStructureManage
 }
 
+VulkanAccelerationStructureManager::Allocation VulkanAccelerationStructureManager::CreateReplayMemory(
+    MemoryType memType, VkDeviceSize size, VkBufferUsageFlags extraUsageFlags)
```
Having a separate path for allocating memory during capture for the copied inputs makes sense for now, but adding a new separate memory allocation path here during replay isn't desirable when we already have `AllocateMemoryForResource` for this exact purpose.

As I can see it the only problem with using that right now is that it doesn't allow freeing memory piecewise so anything allocated with all other upload memory would hang around indefinitely. But that shouldn't be an issue if we add a new `MemoryScope` so that it can be freed in batch once initial contents are uploaded the first time.

cmannett85-arm replied:
Am I OK to add this in a follow-up patch?

baldurk replied:
I'd prefer do that in this PR as the issue I have is adding unnecessary duplicated code - I'd rather not add it at all, than add then remove it. It should be as simple as adding an enum and using the existing functionality.

cmannett85-arm replied:
Yep, no probs.  I've already done the work for using `GPUBuffer` and this new mem scope bit, but the latter is causing failed AS builds so still investigating that one.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -450,320 +618,261 @@ void VulkanAccelerationStructureManager::CopyAccelerationStructure(
-  // not using SERIALISE_ELEMENT_ARRAY so we can deliberately avoid allocation - we serialise
-  // directly into upload memory
-  ser.Serialise("Serialised AS"_lit, contents, contentsSize, SerialiserFlags::NoFlags).Important();
+      VkCommandBuffer cmd = m_pDriver->GetInitStateCmd();
```
`GetInitStateCmd()` is only intended to be used during Prepare and Apply, not during Serialise. I think this might work right now but that's somewhat coincidental and shouldn't be counted on. I think it's better to allocate the memory here and then do the copy with the build on apply (see the comment about caching the AS itself).

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -450,320 +618,261 @@ void VulkanAccelerationStructureManager::CopyAccelerationStructure(
-    m_pDriver->SubmitCmds();
-    m_pDriver->FlushQ();
-  }
+  // We serialise the AS builds so we can have just a single scratch buffer and reuse it
```
Serialising the AS builds should not require a full GPU sync, you should be able to prevent overlap in builds with the same scratch space with a barrier.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -450,320 +618,261 @@ void VulkanAccelerationStructureManager::CopyAccelerationStructure(
-    RDCERR("Couldn't acquire command buffer");
-    return;
-  }
+  const VkAccelerationStructureKHR wrappedAS =
```
I found that repeated AS builds had a pretty big impact on replay time, so I ended up for BLASs building into a temporary scratch GPU-local AS of my own which can then just be copied into the resource being apply'd. The copy ends up much faster than a rebuild.

If you rebase onto latest you'll find that `Apply_InitState` now takes a non-const `VkInitialContents` to allow for this caching and update-in-place.

This would also let you release the uploaded build inputs too (at least for now, in future if we need them for debugging we can keep it around), and in theory the scratch buffer too.

cmannett85-arm replied:
> I ended up for BLASs building into a temporary scratch GPU-local AS of my own

What's the best way of creating this AS?  It needs to be identical to the 'real' AS but AFAICT the creation info has been lost - do I need to capture all the creation info in `VkAccelerationStructureInfo` too or is there a simpler way?

cmannett85-arm replied:
Actually looking at `VkAccelerationStructureCreateInfoKHR` , the only part I don't have is whatever is in the `pNext` chain specifically `VkOpaqueCaptureDescriptorDataCreateInfoEXT`.  No idea if that's a problem.

cmannett85-arm replied:
Ah, I've just seen https://github.com/baldurk/renderdoc/issues/2880.  I guess this approach is fine for now then.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -836,6 +945,162 @@ VulkanAccelerationStructureManager::Allocation VulkanAccelerationStructureManage
+        tri.vertexData.deviceAddress = bufAddr;
+
+        if(tri.indexType != VK_INDEX_TYPE_NONE_KHR)
+          tri.indexData.deviceAddress = bufAddr;
```
I meant to leave a comment here but forgot while jumping around - should the vertex and index data both be pointing to the base of the buffer? shouldn't one be offset if indices are in use?

cmannett85-arm replied:
Assuming I've read the spec right, the offsets are provided by the associated `VkAccelerationStructureBuildRangeInfoKHR`.  So if an index buffer is used it starts reading them from `primitiveOffset` bytes from this address, otherwise it consumes the vertices from this address directly.

baldurk replied:
Ahh I see what you mean, yes you are right there is a (somewhat redundant?) offset internally which is set by CopyInputBuffers and propagate through which covers this.

cmannett85-arm committed:
Reset initial contents after serialisation when softmem limit flushing

If an initial content AS was destroyed during active capture then resetting before serialisation meant that VkInitialContents::Free() was called, deref-ing the AS info and causing it to be deleted.

So VkInitialContents is now reset after serialisation to ensure its lifetime.

Change-Id: I6a296bd96dd5b639f68347073298e5545e8f980d

cmannett85-arm committed:
Use single mem block for AS input data

Rather than have a buffer per geometry, allocate a single block to carry all the input geometry.

Change-Id: Ibc6f221ba0a317e8006ccf4a252da71c0aefe7dc

drbrdt

Change-Id: Ia41dd97fec4f7a2b78fc505dacc0304bbc818789

cmannett85-arm commented:
I've rebased to pick up the mutable `VkInitialContents` change you made, and added the single input data mem block and build-on-apply-once change as separate commits.

cmannett85-arm committed:
Build AS on first Apply()

Then clone to captured AS.  This is much faster than building on each Apply().

baldurk commented on renderdoc/driver/vulkan/vk_initstate.cpp:
```diff
@@ -577,25 +582,22 @@ bool WrappedVulkan::Prepare_InitialState(WrappedVkRes *res)
-    }
+    // Skip empty AS input data (BLASes are force ref-ed)
+    if(record->accelerationStructureInfo->memSize == 0)
+      return true;
```
Is an empty AS invalid, or why do we want to skip that here? I'm assuming you're meaning with the comment that we might be serialising initial states for BLASs that haven't been built.

If we skip serialising here we're not going to rebuild it (unless we add handling for `Create_InitialState`) but I think both an empty BLAS and empty TLAS are degenerate but valid so if you're wanting to check if the AS has been built or not that should be checked for specifically I think, not based on the size of the data.

cmannett85-arm replied:
I'm not sure what I was thinking about when I wrote that comment, as the line actually relates to this one: https://github.com/baldurk/renderdoc/pull/3451/files/e2342e8fa987befcb18292e5c0a2be5088c28643#diff-0fc27a78a09357936e822ace8f29fee4f432a61c7a42113f780104ccbf07911cR539.

I had seen some TLASs that had zero `VkAccelerationStructureBuildGeometryInfoKHR::geometryCount` in a tech demo, so the VVL would kick off about zero sized buffer allocations, copies, etc.  so I added an early exit in `CopyInputBuffers(..)` and ths corresponding one during the prepare stage.  Re-reading the spec, it expressly states that the `geometryCount` must be 1 for a TLAS, but says nothing for a BLAS so I guess it is valid.  I'll update the code so that it skips trying to do anything with empty buffers but otherwise processes it as normal.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -68,6 +118,109 @@ void VkAccelerationStructureInfo::Release()
+        VkDeviceOrHostAddressConstKHR tData;
+        tData.deviceAddress = g.buildRangeInfo.transformOffset
+                                  ? g.memOffset
+                                  : std::numeric_limits<VkDeviceAddress>::max();
```
`VkDeviceAddress` is always 64-bit, this can be replaced with `~0ULL` to avoid the use of `std` and be easier to read.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -24,6 +24,7 @@
 
 #include "vk_acceleration_structure.h"
 #include "core/settings.h"
+#include "limits"
```
(minor note, the header can be removed but including standard headers with `""` is a little weird so I wouldn't recommend it in future at the very least for RenderDoc)

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -68,6 +118,109 @@ void VkAccelerationStructureInfo::Release()
+{
+  uint64_t geomDataSize = 0;
+  for(const GeometryData &geoData : geometryData)
+    geomDataSize += geoData.GetSerialisedSize();
```
I think this would be better simplified as `geoData.byteSize()` which multiplies `sizeof` by count automatically.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -580,183 +613,265 @@ bool VulkanAccelerationStructureManager::Serialise(SerialiserType &ser, Resource
-      const VkDeviceSize size = AlignUp(initial->mem.size, nonCoherentAtomSize);
+    VkAccelerationStructureInfo *asInfo = initial->accelerationStructureInfo;
+    RDCASSERT(asInfo != NULL);
+    SERIALISE_ELEMENT(*asInfo).Hidden();
```
Sorry I should have clarified in my earlier comment - anything which is serialising should be outside of any reading/writing control flow to ensure that the serialisation is easily readable as the same in both directions. The flow of a serialisation function is as below:

```
Data data;
if(IsWriting())
{
  // do anything to prepare data
}

SERIALISE(data);

if(IsReading())
{
process(data);
}
```

For complex cases like initial states here when we're uploading data directly you sometimes need a codepath before the serialisation for reading too, to map the buffer etc. Or you might need to do "two passes" of serialisation, one for the metadata, then allocate a buffer, then serialise the data. In any case the actual call to serialise should be kept outside of control flow.

This is a tradeoff between making it super clear that the serialisation won't mismatch, and the simplification possible by abstracting out as you had before with separate functions.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -580,183 +613,265 @@ bool VulkanAccelerationStructureManager::Serialise(SerialiserType &ser, Resource
-                                         0, (void **)&mappedSrcBuffer);
+    const MemoryAllocation asMemory = m_pDriver->AllocateMemoryForResource(
+        true, mrq, MemoryScope::InitialContents, MemoryType::GPULocal);
+    vkr = m_pDriver->vkBindBufferMemory(d, asBuf, asMemory.mem, asMemory.offs);
```
I think we need to hold onto this `asBuf` and `asMemory` somewhere so we can properly destroy them once we're done with the initial contents at replay shutdown? At the moment we only hold onto `replayAS`, and  destroy the upload buffer/memory.

cmannett85-arm replied:
`asBuf` yes, but won't `asMemory` be cleaned up when all the Initial contents are?

baldurk replied:
Ah yes sorry it will, ignore me reading too quickly :) only `asBuf` indeed.

cmannett85-arm commented on renderdoc/driver/vulkan/vk_core.cpp:
```diff
@@ -2309,6 +2309,7 @@ void WrappedVulkan::StartFrameCapture(DeviceOwnedWindow devWnd)
   // will check to see if they need to markdirty or markpendingdirty
   // and go into the frame record.
   {
+    CheckPendingCommandBufferCallbacks();
```
I've had to move this to just outside the `SCOPED_WRITELOCK(m_CapTransitionLock)` because the readback mem clean up calls `GPUBuffer::Destroy()` which calls into `WrappedVulkan::vkDestroyBuffer` and `vkFreeMemory` - both of which call `SCOPED_READLOCK(m_CapTransitionLock)`.

On bionic this appears to have no ill effects but on glibc it seems to put the lock into a weird state where the readers are still decremented even though they though weren't incremented, as the read count is unsigned this can eventually cause a wrap around.  Then when `SCOPED_WRITELOCK(m_CapTransitionLock)` is called again in `WrappedVulkan::EndFrameCapture` the lock thinks it has readers so waits forever.

baldurk replied:
The RW lock is not recursive I believe so that's why it will go wrong when held recursively.  Moving it here should be fine but can you move it outside of the scope just to be very clear that it's a deliberate choice and not an accidental placement.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -68,11 +111,115 @@ void VkAccelerationStructureInfo::Release()
+{
+  // Scratch is cleaned up automatically, but we need to reset the members for the next run
+  scratch = {};
+  scratchAddressUnion.deviceAddress = 0x0;
```
Not that it's harmful but this comment/function doesn't match up with how it's used - this is resetting values but it only happens right as replay is shutting down so this manager is never re-used after that. Is this meant to be called somewhere else?

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -580,75 +606,36 @@ bool VulkanAccelerationStructureManager::Serialise(SerialiserType &ser, Resource
-  MemoryAllocation uploadMemory;
-  VkBuffer uploadBuf = VK_NULL_HANDLE;
+  SERIALISE_ELEMENT(*asInfo).Hidden();
+  RDCASSERT(asInfo);
```
This assert would be better before the serialise as otherwise if it's ever NULL we will crash before getting to the assert.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.cpp:
```diff
@@ -766,74 +866,142 @@ void VulkanAccelerationStructureManager::Apply(ResourceId id, const VkInitialCon
+  return result;
+}
+
+VulkanAccelerationStructureManager::Allocation VulkanAccelerationStructureManager::CreateReplayMemory(
```
I don't mind leaving this as a helper but I think this function (honestly both functions) would be better named as `CreateReplayBuffer` or `CreateTempReplayBuffer` as it's not just allocating memory.

cmannett85-arm committed:
New memory scope that clears on after first Apply()

Also:
* Use GPUBuffer instead of own radback memory mechanism, but mark as internal
* Use GPU local buffer for AS builds instead of upload memory to improve performance
